### PR TITLE
[Element 2.0] Implement multiple test data from element 1.5 to 2.0

### PR DESCRIPTION
**Pull request by honglax**
Based on [this PR](https://github.com/flood-io/element/pull/308)

### CODE REVIEW

File: packages/core/src/runtime/EvaluatedScript.ts [hunk 1]
```diff
@@ -121,14 +121,14 @@ export class EvaluatedScript implements TestScriptErrorMapper, EvaluatedScriptLi
...
-		await this.testData.load()
+		if (this.testData) await this.testData.load()
 	}
 
 	private _testDataLoaders: TestDataFactory | undefined
 	public get testDataLoaders(): TestDataFactory {
 		if (this._testDataLoaders === undefined) {
 			this._testDataLoaders = new BoundTestDataLoaders(this, this.runEnv.workRoot)
-			this._testDataLoaders.fromData([{}])
+			// this._testDataLoaders.fromData([{}])
```
Comment (phileba) [inline packages/core/src/runtime/EvaluatedScript.ts:131]: should remove this line

File: packages/core/src/runtime/Test.ts [hunk 2]
```diff
@@ -180,9 +179,32 @@ export default class Test implements ITest {
 			await testObserver.before(this)
 
 			debug('Feeding data')
-			testDataRecord = testData.feed()
-			if (testDataRecord === null) {
-				throw new Error('Test data exhausted, consider making it circular?')
+			// testDataRecord = testData.feed()
+			// if (testDataRecord === null) {
+			// 	throw new Error('Test data exhausted, consider making it circular?')
```
Comment (phileba) [inline packages/core/src/runtime/Test.ts:184]: the same above

**Review by phileba — APPROVED** [review]

File: examples/using-test-data/multi-test-data.perf.ts [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	screenshotOnFailure: false,
+	clearCache: true,
+	clearCookies: true,
+	disableCache: true,
+	incognito: true,
+	actionDelay: '1s',
+	stepDelay: '2s',
+}
+
+type Flooders = {
```
Comment (HieuLeKMS) [inline examples/using-test-data/multi-test-data.perf.ts:16]:
```suggestion
type Flooder = {
```

File: packages/core/src/test-data/Loader.ts [hunk 4]
```diff
@@ -67,56 +145,96 @@ export class DataLoader<T> extends Loader<T> {
...
-		const data = readFilePromise(this.filePath, 'utf8')
-		data.catch(err => {
-			console.error(err)
-		})
-
-		const jsonData: T[] = JSON.parse(await data)
-		if (Array.isArray(jsonData)) {
-			this.lines = jsonData
+		if (this.filePaths.length) {
+			const allLines: any[] = []
```
Comment (HieuLeKMS) [inline packages/core/src/test-data/Loader.ts:160]: I could be wrong but I think it should be like this
```suggestion
			const allLines: T[] = []
```

File: packages/core/src/test-data/Loader.ts [hunk 5]
```diff
@@ -67,56 +145,96 @@ export class DataLoader<T> extends Loader<T> {
...
-			data = await readFilePromise(this.filePath, 'utf8')
-		} catch (e) {
-			throw new Error(`unable to read CSV file ${this.filePath}:\ncause: ${e}`)
+		const option = {
+			type: FileType.CSV,
+			delimiter: this.separator,
+			columns: true,
+		}
+		if (this.filePaths.length) {
+			const allLines: any[] = []
```
Comment (HieuLeKMS) [inline packages/core/src/test-data/Loader.ts:206]: Same as my comment above, please just change nothing and resolve this if I'm wrong
```suggestion
			const allLines: T[] = []
```

**Review by HieuLeKMS — APPROVED** [review]

File: packages/core/src/runtime/EvaluatedScript.ts (hunk 1 above)
Comment (honglax) [reply packages/core/src/runtime/EvaluatedScript.ts:131]: Resolved

File: packages/core/src/runtime/Test.ts (hunk 2 above)
Comment (honglax) [reply packages/core/src/runtime/Test.ts:184]: Resolved

File: examples/using-test-data/multi-test-data.perf.ts (hunk 3 above)
Comment (honglax) [reply examples/using-test-data/multi-test-data.perf.ts:16]: Resolved

File: packages/core/src/test-data/Loader.ts (hunk 4 above)
Comment (honglax) [reply packages/core/src/test-data/Loader.ts:160]: Resolved

File: packages/core/src/test-data/Loader.ts (hunk 5 above)
Comment (honglax) [reply packages/core/src/test-data/Loader.ts:206]: Resolved

**Pull request merged**